addons/binding/org.openhab.binding.amazonechocontrol/src/main/java/org/openhab/binding/amazonechocontrol/internal/statedescription/AmazonEchoDynamicStateDescriptionProvider.java

```diff
@@ -0,0 +1,290 @@
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Dynamic channel state description provider.
+ * Overrides the state description for the controls, which receive its configuration in the runtime.
+ *
+ * @author Michael Geramb - Initial contribution
+ */
+@NonNullByDefault
+@Component(service = { DynamicStateDescriptionProvider.class,
+        AmazonEchoDynamicStateDescriptionProvider.class }, immediate = true)
+public class AmazonEchoDynamicStateDescriptionProvider implements DynamicStateDescriptionProvider {
+
+    private final Logger logger = LoggerFactory.getLogger(AmazonEchoDynamicStateDescriptionProvider.class);
+    private @Nullable ThingRegistry thingRegistry;
+
+    @Reference(cardinality = ReferenceCardinality.MANDATORY, policy = ReferencePolicy.DYNAMIC)
+    protected void setThingRegistry(ThingRegistry thingRegistry) {
+        this.thingRegistry = thingRegistry;
+    }
+
+    protected void unsetThingRegistry(ThingRegistry thingRegistry) {
+        this.thingRegistry = thingRegistry;
+    }
+
+    public @Nullable ThingHandler findHandler(Channel channel) {
+        ThingRegistry thingRegistry = this.thingRegistry;
+        if (thingRegistry == null) {
+            return null;
+        }
+        Thing thing = thingRegistry.get(channel.getUID().getThingUID());
+        if (thing == null) {
+            return null;
+        }
+        return thing.getHandler();
+    }
+
+    StateOption CreateStateOption(@Nullable String id, @Nullable String displayValue, boolean showIdsInGUI) {
+        if (showIdsInGUI) {
+            return new StateOption(id, String.format("%s [%s]", displayValue, id));
+        } else {
+            return new StateOption(id, displayValue);
+        }
+    }
+
+    @Override
+    public @Nullable StateDescription getStateDescription(Channel channel,
+            @Nullable StateDescription originalStateDescription, @Nullable Locale locale) {
+        if (originalStateDescription == null) {
+            return null;
+        }
+        ThingRegistry thingRegistry = this.thingRegistry;
+        if (thingRegistry == null) {
+            return originalStateDescription;
+        }
+        if (CHANNEL_TYPE_BLUETHOOTH_MAC.equals(channel.getChannelTypeUID())) {
+            EchoHandler handler = (EchoHandler) findHandler(channel);
+            if (handler == null) {
+                return originalStateDescription;
+            }
+            BluetoothState bluetoothState = handler.findBluetoothState();
+            if (bluetoothState == null) {
+                return originalStateDescription;
+            }
+            PairedDevice[] pairedDeviceList = bluetoothState.pairedDeviceList;
+            if (pairedDeviceList == null) {
+                return originalStateDescription;
+            }
+
+            ArrayList<StateOption> options = new ArrayList<StateOption>();
```

Comment:
You should use diamond operators (`new ArrayList<>();`) all throughout the code.